### ISSUE TITLE
Fix layer_norm_backward launch dimensions for 3-D inputs

### ISSUE THREAD

**Issue by lvyufeng** [FlagOpen/FlagGems#5490]
## Summary

`flag_gems.ops.layernorm.layer_norm_backward` assumes a 2-D input when computing its Triton launch dimensions. For a normal transformer input with shape `[batch, sequence, hidden]` and `normalized_shape=(hidden,)`, it computes `M` from `input.shape[0]` instead of the product of `normalized_shape`. This produces incorrect gradients and can make the weight/bias gradient kernel write out of bounds, corrupting the device allocator and poisoning the MetaX runtime for the rest of the process.

## Environment

- Hardware: MetaX C550
- Driver/runtime: MACA 3.8.x
- Python: 3.12
- PyTorch: 2.10.0+cpu with torch_fl MetaX CUDA-ABI boxing
- FlagGems: `5.4.0.dev20260810+g7fb49bad4` (the current checkout also has the same code as upstream `master`)
- Transformers: v5.12.1 workload shapes

Runtime configuration:

```bash
FLAGOS_METAX_BOXING=1 FLAGOS_USE_FLAGGEMS=1
```

## Reproduction

```python
import torch_fl  # Must be imported before torch for MetaX boxing.
import torch

torch.manual_seed(0)
M, L, H = 52, 7, 64
x_cpu = torch.randn(M, L, H)
w_cpu = torch.randn(H)
b_cpu = torch.randn(H)
g_cpu = torch.randn(M, L, H)

x = x_cpu.to("flagos").requires_grad_(True)
w = w_cpu.to("flagos").requires_grad_(True)
b = b_cpu.to("flagos").requires_grad_(True)
torch.nn.functional.layer_norm(x, (H,), w, b).backward(g_cpu.to("flagos"))
torch.flagos.synchronize()

xr = x_cpu.clone().requires_grad_(True)
wr = w_cpu.clone().requires_grad_(True)
br = b_cpu.clone().requires_grad_(True)
torch.nn.functional.layer_norm(xr, (H,), wr, br).backward(g_cpu)

print("gx maxdiff:", float((x.grad.cpu() - xr.grad).abs().max()))
print("gw maxdiff:", float((w.grad.cpu() - wr.grad).abs().max()))
print("gb maxdiff:", float((b.grad.cpu() - br.grad).abs().max()))
```

Run it with the FlagGems route enabled:

```bash
FLAGOS_METAX_BOXING=1 FLAGOS_USE_FLAGGEMS=1 python repro_ln_bwd.py
```

## Expected behavior

The FlagGems result should match the CPU reference up to normal floating-point tolerance (approximately `1e-5` for float32).

## Actual behavior

With FlagGems enabled, the reproduced differences were:

```text
gx maxdiff: 1.8476690299629362e+32
gw maxdiff: 48.97093200683594
gb maxdiff: 47.84790802001953
```

The same script using the CUDA boxing route produced:

```text
gx maxdiff: 9.5367431640625e-07
gw maxdiff: 9.5367431640625e-06
gb maxdiff: 1.1444091796875e-05
```

Larger shapes can trigger a MetaX ATU/Xnack fault in `layer_norm_backward_kernel`; after that fault, subsequent allocations report `CachingDeviceAllocator: invalid device index` because the MACA runtime has been disabled at process scope.

## Root cause

In `src/flag_gems/ops/layernorm.py`, the forward path correctly computes:

```python
N = math.prod(normalized_shape)
M = input.numel() // N
```

but `layer_norm_backward` currently computes:

```python
M = input.shape[0]
N = input.numel() // M
```

For `[52, 7, 64]`, this changes the intended `(M, N) = (364, 64)` into `(52, 448)`. The `weight_bias_backward_kernel` then launches over 448 columns while `weight_grad` and `bias_grad` each contain only 64 elements, allowing masked stores past the end of both output tensors. Even when this does not immediately fault, the gradients are numerically wrong.

## Proposed fix

Use the same flattening as the forward implementation:

```python
N = math.prod(normalized_shape)
M = input.numel() // N
```

Please also add a regression test covering at least a 3-D input with `normalized_shape` equal to the final dimension, and compare `dx`, `dweight`, and `dbias` against PyTorch.

## Investigation

- Read `src/flag_gems/ops/layernorm.py` in both the current FlagGems checkout and upstream `master`.
- Compared forward and backward launch-dimension calculations.
- Reproduced the numerical failure on MetaX with the FlagGems route.
- Verified that routing `native_layer_norm_backward` to the CUDA boxing kernel restores CPU-level numerical agreement.

This issue is filed from the torch_fl MetaX integration investigation. A temporary downstream fallback to the CUDA boxing path is being used until this upstream fix is available.
